### PR TITLE
Add BGC fields exchanged between ice and ocn to the cpl carbon budget

### PR DESCRIPTION
**Pull request by jonbob**
Adds BGC fields that are exchanged between the seaice and ocean to the coupler carbon budget. The fields include three different algae, dissolved inorganic carbon and four dissolved organic carbon fluxes. The results in the coupler budget match with the output from the seaice carbon conservation analysis and are balanced in the E3SM system.

[BFB]

### CODE REVIEW

**Comment by jonbob** [comment]
Sample output from the budget after a six-month test using ne30pg2_r05_EC30to60E2r2 and the BGCEXP_CNTL_CNPECACNT_1850 compset:
````
(seq_diagBGC_print_mct) NET CARBON BUDGET (kg-C/m2s*1e10): period =  monthly: date =     10701     0
                           atm            lnd            rof            ocn         ice nh         ice sh            glc        *SUM*
     surface co2   -12.09362044    11.56416670     0.00000000     0.52945373     0.00000000     0.00000000     0.00000000    -0.00000000
    black carbon    -0.00086318     0.00047700     0.00000000     0.00036538     0.00001977     0.00000097     0.00000000    -0.00000007
    orgnc carbon    -0.03569801     0.01656353     0.00000000     0.01750498     0.00149243     0.00012880     0.00000000    -0.00000826
     i-o inorgnc     0.00000000     0.00000000     0.00000000     0.91426593    -0.55158401    -0.36308742     0.00000000    -0.00040551
     i-o   orgnc     0.00000000     0.00000000     0.00000000    -0.23069659    -0.30029178     0.53079940     0.00000000    -0.00018897
           *SUM*   -12.13018163    11.58120724     0.00000000     1.23089343    -0.85036360     0.16784174     0.00000000    -0.00060281
````
Output from the mpas-seaice carbon conservation am agrees well:
````
 Individual ocean carbon fluxes     :
  Ocean diatom flux       (kg/m2s) =    4.032854E-12   3.899167E-12   1.336867E-13
  Ocean small algae flux  (kg/m2s) =    7.554319E-14   2.023166E-14   5.531153E-14
  Ocean polysaccharid flux(kg/m2s) =   -8.828458E-12   1.118841E-12  -9.947299E-12
  Ocean lipid flux        (kg/m2s) =   -5.573669E-12   8.946094E-13  -6.468279E-12
  Ocean DIC flux          (kg/m2s) =    9.146714E-11   5.515840E-11   3.630874E-11
  Ocean DON flux          (kg/m2s) =   -2.712383E-11  -1.603647E-12  -2.552018E-11
  Ocean humics flux       (kg/m2s) =    1.436680E-11   2.569997E-11  -1.133318E-11
  Sum of ocean fluxes     (kg/m2s) =    6.841638E-11   8.518758E-11  -1.677120E-11
````
where the "Ocean DIC flux" is directly comparable to the coupler "i-o inorgnc". The other comparison requires adding the remaining organic fluxes and comparing to the "i-o orgnc" line: which in mpassi gives -2.305076E-11 compared to -2.3069659E-11 in the coupler.

File: driver-mct/main/seq_diagBGC_mct.F90 [hunk 1]
```diff
@@ -932,8 +970,16 @@ subroutine seq_diagBGC_ice_mct( ice, frac_i, infodata, do_i2x, do_x2i)
     ko    = mct_aVect_indexRA(frac_i,ofracname)
 
     if (present(do_i2x)) then
-!       if (first_time) then
-!       endif
+       if (first_time) then
```
Comment (bpbond) [inline driver-mct/main/seq_diagBGC_mct.F90:973]: Is it worth a comment to explain why handling the first time differently?

File: components/mpas-seaice/driver/ice_comp_mct.F [hunk 2]
```diff
@@ -1704,7 +1704,7 @@ subroutine ice_import_mct(x2i_i, errorCode)!{{{
 !    o  bcphiwet -- Black Carbon hydrophilic wet deposition flux
 !    o  ocphidry -- Organic Carbon hydrophilic dry deposition flux
 !    o  ocphodry -- Organic Carbon hydrophobic dry deposition flux
-!    o  ocphiwet -- Organic Carbon hydrophilic dry deposition flux
+!    o  ocphiwet -- Organic Carbon hydrophilic wet deposition flux
```
Comment (bpbond) [inline components/mpas-seaice/driver/ice_comp_mct.F:1707]: Good catch

**Review by bpbond — APPROVED** [review]
Thanks @jonbob — looks good. Nicely documented!

**Comment by njeffery** [comment]
@jonbob :  This looks good to me.  

Could you test your analysis member with this change in ice_comp_mct?
i2x_i % rAttr(index_i2x_Fioi_algae3,n) = oceanAlgaeFlux(3,i) * carbonToNitrogenRatioAlgae(3)
to
i2x_i % rAttr(index_i2x_Fioi_algae3,n) = 0 ! oceanAlgaeFlux(3,i) * carbonToNitrogenRatioAlgae(3)

Just to verify that oceanAlgaeFlux(3,i) is zero.  thanks

File: driver-mct/main/seq_diagBGC_mct.F90 (hunk 1 above)
Comment (jonbob) [reply driver-mct/main/seq_diagBGC_mct.F90:973]: Always worth a comment! Throughout the budget codes we use "first_time" as a logic wrapper to set indices used over the course of the run that only need to be set once. Previously there were no BGC fields included in the carbon from the ice to the cpl (i2x), so the logic was commented out. This PR adds the fluxes that are exchanged between the ice and ocn through the cpl, so now it is needed. I'll add a short description of first_time to the code to make that more clear.

**Review by njeffery — APPROVED** [review]
Approved. Verification is based on visual inspection.

**Comment by jonbob** [comment]
@njeffery - I did rerun the test with 
````
i2x_i % rAttr(index_i2x_Fioi_algae3,n) = 0
````
and got exactly the same results

**Comment by njeffery** [comment]
@jonbob : Excellent! thanks.

**Comment by jonbob** [comment]
@jenniferholm - it's not necessary to review this PR, but I wanted to give you the chance if you have time

**Comment by jonbob** [comment]
Passes:
* SMS_Ld2.ne30pg2_r05_EC30to60E2r2.BGCEXP_CNTL_CNPECACNT_1850.chrysalis_intel.elm-bgcexp
* ERS.ne11_oQU240.WCYCL1850NS.anvil_intel

merged to next

**Comment by jonbob** [comment]
merged to master

**Pull request merged**